src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -519,22 +556,27 @@ private static Type reify(final Type genericType, final Map<TypeVariable<?>, Typ
     } else if (genericType instanceof TypeVariable<?>) {
       final TypeVariable<?> typeVariable = (TypeVariable<?>) genericType;
       final Type mapping = typeVariableMap.get(typeVariable);
-      if (mapping != null)
-        return reify(mapping, typeVariableMap);
+      if (mapping != null) {
+        cache = new HashMap<>(cache);
+        cache.put(typeVariable, mapping);
+        return reify(mapping, typeVariableMap, cache);
+      }
 
       final Type[] upperBounds = typeVariable.getBounds();
       // NOTE: According to https://docs.oracle.com/javase/tutorial/java/generics/bounded.html
       // if there are multiple upper bounds where one bound is a class, then this must be the
       // leftmost/first bound. Therefore we blindly take this one, hoping is the most relevant.
       // Hibernate does the same when erasing types, see also
       // https://github.com/hibernate/hibernate-validator/blob/6.0/engine/src/main/java/org/hibernate/validator/internal/util/TypeHelper.java#L181-L186
-      return reify(upperBounds[0], typeVariableMap);
+      cache = new HashMap<>(cache);
+      cache.put(typeVariable, upperBounds[0]);
```

Comment:
Could you describe a bit about why we create a new cache here, and also a few lines above?